src/bsoncxx/types/value.cpp

```diff
@@ -56,35 +78,57 @@ value& value::operator=(const value& rhs) {
 
 value::value(value&& rhs) noexcept {
     *this = std::move(rhs);
+    rhs._type = stdx::nullopt;
 }
 
 value& value::operator=(value&& rhs) noexcept {
-    _type = rhs._type;
+    if (this == &rhs) {
+        // TODO: use bsoncxx error category
+        throw std::runtime_error("self move assign");
+    }
 
-    switch (static_cast<int>(_type)) {
-#define BSONCXX_ENUM(type, val)               \
-    case val:                                 \
-        _b_##type = std::move(rhs._b_##type); \
-        break;
+    if (_type) {
+            switch (static_cast<typename std::underlying_type<bsoncxx::type>::type>(*_type)) {
+#define BSONCXX_ENUM(type, val)     \
+        case val:                   \
+            _b_##type.~b_##type();  \
+            break;
 #include <bsoncxx/enums/type.hpp>
 #undef BSONCXX_ENUM
+        }
+    }
+
+    _type = rhs._type;
+    rhs._type = stdx::nullopt;
```

Comment:
Rather than this, after the code below where you move from rhs, add a `rhs.destroy()` call.